api/agent/agent.go

```diff
@@ -278,55 +284,95 @@ func (a *agent) getSlot(ctx context.Context, call *call) (Slot, error) {
+	timeout := time.Duration(60) * time.Minute
+	idleTimeout := time.Duration(call.IdleTimeout) * time.Second * 2
+	if timeout < idleTimeout {
+		timeout = idleTimeout
+	}
+
+	logger := logrus.WithFields(logrus.Fields{
+		"app":              call.AppName,
+		"route":            call.Path,
+		"image":            call.Image,
+		"memory":           call.Memory,
+		"format":           call.Format,
+		"idle_timeout":     call.IdleTimeout,
+		"launcher_timeout": timeout})
 
+	logger.Info("Hot function launcher starting")
 	isAsync := call.Type == models.TypeAsync
 
-launchLoop:
+loop:
 	for {
-		// Check/evaluate if we need to launch a new hot container
-		doLaunch, stats := call.slots.isNewContainerNeeded()
-		common.Logger(ctx).WithField("stats", stats).Debug("checking hot container launch ", doLaunch)
+		select {
+		case <-a.shutdown: // server shutdown
+			break loop
+		case <-time.After(timeout):
+			if call.slots.isIdle() {
+				logger.Info("Hot function launcher timed out")
+				break loop
+			}
+		case <-call.slots.signaller:
+		}
 
-		if doLaunch {
-			ctxToken, tokenCancel := context.WithCancel(context.Background())
+		isNeeded, stats := call.slots.isNewContainerNeeded()
+		logger.WithField("stats", stats).Debug("Hot function launcher stats")
+		if !isNeeded {
+			continue loop
+		}
 
-			// wait on token/slot/timeout whichever comes first
-			select {
-			case tok, isOpen := <-a.resources.GetResourceToken(ctxToken, call.Memory, isAsync):
-				tokenCancel()
-				if !isOpen {
-					return nil, models.ErrCallTimeoutServerBusy
-				}
-				go a.runHot(ctx, call, tok)
-			case s, ok := <-call.slots.getDequeueChan():
-				tokenCancel()
-				if !ok {
-					return nil, errors.New("slot shut down while waiting for hot slot")
-				}
-				if s.acquireSlot() {
-					if s.slot.Error() != nil {
-						s.slot.Close()
-						return nil, s.slot.Error()
-					}
-					return s.slot, nil
-				}
+		resourceCtx, cancel := context.WithCancel(context.Background())
+		logger.WithField("stats", stats).Info("Hot function launcher starting hot container")
 
-				// we failed to take ownership of the token (eg. container idle timeout)
-				// try launching again
-				continue launchLoop
-			case <-ctx.Done():
-				tokenCancel()
-				return nil, ctx.Err()
+		select {
+		case tok, isOpen := <-a.resources.GetResourceToken(resourceCtx, call.Memory, isAsync):
+			cancel()
+			if isOpen {
+				go a.runHot(ctx, call, tok)
+			} else {
+				// this means the resource was impossible to reserve (eg. memory size we can never satisfy)
+				call.slots.queueSlot(&hotSlot{done: make(chan struct{}), err: models.ErrCallTimeoutServerBusy})
 			}
+		case <-time.After(timeout):
+			cancel()
+			if call.slots.isIdle() {
+				logger.Info("Hot function launcher timed out")
+				break loop
+			}
+		case <-a.shutdown: // server shutdown
+			cancel()
+			break loop
+		}
+	}
+
+	logger.Info("Hot function launcher terminating")
+	a.slotMgr.destroySlotQueue(call.slots)
+	a.wg.Done()
```

Comment:
shouldn't we call this with `defer`?

Reply:
I fail to see why defer is preferable here. Code layout is clear and there's one exit path out of the function.

Reply:
Note that defer is also slower.

Reply:
It really looks like there's only one (i think) exit path then yes, defer is redundant

Reply:
So, i'm not sure for how slow defer would be here, but i've seen reports that defer with anonymous function works faster:
```go
defer func(){ a.wg.Done() }()
```